### ISSUE TITLE
[BUG] pyarmor.cli.CliError: could not get docker host machine id

### ISSUE THREAD

**Issue by broall** [dashingsoft/pyarmor#1407]
Hi,

I'm trying to register group license in my machine. My machine is Mac M1, os version 13.4.
Here is the result for running command `pyarmor gen`
```
$ pyarmor gen -O dist --assert-call foo.py
INFO     Python 3.9.12
INFO     Pyarmor 8.3.0 (trial), 000000, non-profits
INFO     Platform darwin.x86_64
ERROR    no found input "foo.py"
```

Actually, I'm following the guideline [2.9.3.5. Run unlimited dockers in offline device](https://pyarmor.readthedocs.io/en/stable/how-to/register.html#contents). I could successfully generate file `pyarmor-device-regfile-6065.1.zip` from `pyarmor-regfile-6065.zip` file. 

Now, I run this command and get error:
```
$ pyarmor-auth pyarmor-device-regfile-6065.1.zip
could not get docker host machine id
Traceback (most recent call last):
  File "/opt/anaconda3/lib/python3.9/site-packages/pyarmor/cli/register.py", line 183, in _get_docker_hostname
    s.connect((host, port))
ConnectionRefusedError: [Errno 61] Connection refused
Traceback (most recent call last):
  File "/opt/anaconda3/bin/pyarmor-auth", line 8, in <module>
    sys.exit(main())
  File "/opt/anaconda3/lib/python3.9/site-packages/pyarmor/cli/docker.py", line 107, in main
    main_entry()
  File "/opt/anaconda3/lib/python3.9/site-packages/pyarmor/cli/docker.py", line 90, in main_entry
    register_pyarmor(ctx, args.regfile[0])
  File "/opt/anaconda3/lib/python3.9/site-packages/pyarmor/cli/docker.py", line 62, in register_pyarmor
    reg.register_regfile(regfile)
  File "/opt/anaconda3/lib/python3.9/site-packages/pyarmor/cli/register.py", line 161, in register_regfile
    raise CliError('could not get docker host machine id')
pyarmor.cli.CliError: could not get docker host machine id
```

It seems that the script will try to connect to `host.docker.internal`, which is unavailable in Mac. I tested on ubuntu 22 and that DNS is available.

Please guide me how to register group license in my machine as I have been stucked for several days.

**Comment by jondy** [comment]
This Mac M1 is docker container or physical machine? If it's docker container, you should not run `pyarmor-auth` on it. `pyarmor-auth` is running in docker host.

If this Mac M1 is docker host, check machine id to make sure it doesn't changed.

This is the machine id when generating group license
```
cat .pyarmor/group/pyarmor-group-device.1
```

And this is machine id now
```
python
>>> from pyarmor.cli.core import Pytransform3
>>> print(Pytransform3.get_hd_info(16))
```

If they're same, could you show the steps in details?

**Comment by broall** [comment]
Thank you for your reply. Yes, mac is a physical machine, I already installed docker in it.

Here are those results when I ran those commands:

```
$ cat .pyarmor/group/pyarmor-group-device.1
# Generated by Pyarmor 8.3.0, 2023-08-01T22:29:25.252901
host: 192.168.2.4
system: Darwin (Darwin Kernel Version 22.5.0: Mon Apr 24 20:52:24 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T6000)
machine: g174f50205aef4868fbc4f89213b4ac90%
```

```
$ python
Python 3.9.12 (main, Apr  5 2022, 01:53:17) 
[Clang 12.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyarmor.cli.core import Pytransform3
>>> print(Pytransform3.get_hd_info(16))
b'g174f50205aef4868fbc4f89213b4ac90'
```

For those steps I have done, first I ran this to generate license
```
pyarmor reg -p XXX pyarmor-regcode-6065.txt
```

New registration file `pyarmor-regfile-6065.zip` was generated. Next, still on my Mac, I ran this:
```
pyarmor reg -g 1
pyarmor reg -g 1 /path/to/pyarmor-regfile-6065.zip
```

After ran those commands, `pyarmor-device-regfile-6065.1.zip` was generated. However, when I ran `pyarmor -v`, it still showed that my pyarmor is in trial.

Then I ran this command, and get the error above.
```
pyarmor-auth pyarmor-device-regfile-6065.1.zip
```

**Comment by jondy** [comment]
In the `.pyarmor/group/pyarmor-group-device.1`, the last line `machine: g174f50205aef4868fbc4f89213b4ac90%`
There is extra `%`, please make sure there is no `%` in the last line by `vi .pyarmor/group/pyarmor-group-device.1`

If there is `%` in this device file, this is the problem. Please generate another device with `-g 2` in this machine and try again.

If there is no `%`, try to register in this machine
```
pyarmor reg pyarmor-device-regfile-6065.1.zip
pyarmor -v
```

**Comment by broall** [comment]
Thank you.
When I checked, there is no `%`:
```
$ vi .pyarmor/group/pyarmor-group-device.1
# Generated by Pyarmor 8.3.0, 2023-08-01T22:29:25.252901
host: 192.168.2.4
system: Darwin (Darwin Kernel Version 22.5.0: Mon Apr 24 20:52:24 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T6000)
machine: g174f50205aef4868fbc4f89213b4ac90
```
I reran this command and get error:
```
$ pyarmor reg pyarmor-device-regfile-6065.1.zip
INFO     Python 3.9.12
INFO     Pyarmor 8.3.0 (trial), 000000, non-profits
INFO     Platform darwin.x86_64
INFO     register "pyarmor-device-regfile-6065.1.zip"
ERROR    could not get docker host machine id
Traceback (most recent call last):
  File "/opt/anaconda3/lib/python3.9/site-packages/pyarmor/cli/register.py", line 183, in _get_docker_hostname
    s.connect((host, port))
ConnectionRefusedError: [Errno 61] Connection refused
ERROR    could not get docker host machine id
```

**Comment by jondy** [comment]
Do you notice the machined id is different?

In the `.pyarmor/group/pyarmor-group-device.1`, the last line `machine: g174f50205aef4868fbc4f89213b4ac90%`
There is extra `%`, please make sure there is no `%` in the last line by `vi .pyarmor/group/pyarmor-group-device.1`

If there is `%` in this device file, this is the problem. Please generate another device with `-g 2` in this machine and try again.

**Comment by broall** [comment]
I registered new one and the result is still the same:

```
$ pyarmor reg -g 2
INFO     Python 3.9.12
INFO     Pyarmor 8.3.0 (trial), 000000, non-profits
INFO     Platform darwin.x86_64
INFO     generating device file ".pyarmor/group/pyarmor-group-device.2"
INFO     device file has been generated successfully
```

```
cat .pyarmor/group/pyarmor-group-device.2
# Generated by Pyarmor 8.3.0, 2023-08-03T17:07:36.813969
host: 192.168.2.4
system: Darwin (Darwin Kernel Version 22.5.0: Mon Apr 24 20:52:24 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T6000)
machine: g174f50205aef4868fbc4f89213b4ac90% 
```

The `%` is shown in terminal with red. I think that's a special character, not the normal `%`

**Comment by jondy** [comment]
Reproduced, it will be fixed in next release.

**Comment by broall** [comment]
Thank you so much

**Comment by jondy** [comment]
As a quick workaround, apply this patch
```
--- a/pyarmor/cli/register.py
+++ b/pyarmor/cli/register.py
@@ -153,7 +153,7 @@ class Register(object):
                 f.extract(item, path=path)
             namelist = f.namelist()
             if 'group.tokens' in namelist:
-                machid = self._get_machine_id().decode('utf-8')
+                machid = self._get_machine_id(16).decode('utf-8')
                 name = '/'.join(['tokens', machid])
                 if name not in namelist:
                     machid = self._get_docker_hostname()
```

**Comment by broall** [comment]
The workaround works now. Thank you so much

**Comment by jondy** [comment]
Fixed in 8.3.1
